commit/internal/builder/report.go

```diff
@@ -0,0 +1,396 @@
+			)
+			if err != nil {
+				return nil, err
+			}
+			if report != nil {
+				reports = append(reports, *report)
+			}
+
+			// reset accumulators for next report.
+			numRoots = 0
+			blessedMerkleRoots = make([]cciptypes.MerkleRootChain, 0)
+			unblessedMerkleRoots = make([]cciptypes.MerkleRootChain, 0)
+
+			// price updates are only included in the first report.
+			priceUpdates = cciptypes.PriceUpdates{}
+		}
+	}
+
+	// check for final partial report.
+	if numRoots > 0 {
+		report, err := buildOneReport(
+			ctx,
+			lggr,
+			reportCodec,
+			transmissionSchedule,
+			outcome.MerkleRootOutcome.OutcomeType,
+			blessedMerkleRoots,
+			unblessedMerkleRoots,
+			nil, // no RMN for partial reports.
+			0,   // no RMN for partial reports.
+			priceUpdates,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if report != nil {
+			reports = append(reports, *report)
+		}
+	}
+
+	return reports, nil
+}
+
+// buildMultiplePriceReports is a helper to splits up price data into multiple reports.
+// Helper for buildMultiplePriceAndMerkleRootReports.
+// Merkle root data is ignored.
+func buildMultiplePriceReports(
+	ctx context.Context,
+	lggr logger.Logger,
+	reportCodec cciptypes.CommitPluginCodec,
+	transmissionSchedule *ocr3types.TransmissionSchedule,
+	outcome committypes.Outcome,
+	maxPricesPerReport uint64, // passed in directly to avoid implementing ReportBuilderFunc
+) ([]ocr3types.ReportPlus[[]byte], error) {
+	// update joins together the different types of price updates so that they can be
+	// selected in one loop.
+	type update struct {
+		cciptypes.TokenPrice
+		cciptypes.GasPriceChain
+	}
+	// Helper to add an update to the PriceUpdates type.
+	add := func(updates cciptypes.PriceUpdates, update update) cciptypes.PriceUpdates {
+		if (update.TokenPrice != cciptypes.TokenPrice{}) {
+			updates.TokenPriceUpdates = append(updates.TokenPriceUpdates, update.TokenPrice)
+		}
+		if (update.GasPriceChain != cciptypes.GasPriceChain{}) {
+			updates.GasPriceUpdates = append(updates.GasPriceUpdates, update.GasPriceChain)
+		}
+		return updates
+	}
+
+	var updates []update
+	for _, tokenPriceUpdate := range outcome.TokenPriceOutcome.TokenPrices.ToSortedSlice() {
+		updates = append(updates, update{
+			TokenPrice: tokenPriceUpdate,
+		})
+	}
+	for _, gasPriceUpdate := range outcome.ChainFeeOutcome.GasPrices {
+		updates = append(updates, update{
+			GasPriceChain: gasPriceUpdate,
+		})
+	}
+
+	var reports []ocr3types.ReportPlus[[]byte]
+	numUpdates := uint64(0)
+	priceUpdates := cciptypes.PriceUpdates{}
+
+	for _, u := range updates {
+		numUpdates++
+		priceUpdates = add(priceUpdates, u)
+
+		if numUpdates == maxPricesPerReport {
+			report, err := buildOneReport(
+				ctx,
+				lggr,
+				reportCodec,
+				transmissionSchedule,
+				outcome.MerkleRootOutcome.OutcomeType,
+				nil,
+				nil,
+				nil, // no RMN for partial reports.
+				0,   // no RMN for partial reports.
+				priceUpdates,
+			)
+			if err != nil {
+				return nil, err
```

Comment:
Would more ctx added to the error be useful?

```suggestion
				return nil, fmt.Errorf("building report with only price updates %+v: %w", priceUpdates, err)
```